### PR TITLE
core: Add support for multi-threaded MPIDR values

### PR DESCRIPTION
**Pull request by imre-kis-arm**
If the MT bit is set the affinities are shifted in the MPIDR register
so the get_core_pos_mpidr functions needs to be modified accordingly.
This is necessary to make OP-TEE to be able to run on multi-threaded
systems. The number of threads/core can be modified by the
CFG_THREAD_CORE_SHIFT makefile parameter. The default value is the
existing single threaded mode.

Change-Id: Ibfacad34ac2bc67b1d0b7929f731da6406a97b5a
Signed-off-by: Imre Kis <imre.kis@arm.com>

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
What determines which value `CFG_THREAD_CORE_SHIFT` should have? Is it something that is hardwired into the CPU or can it be configured in some way?

**Comment by jforissier** [comment]
Any chance it could be detected at runtime?

**Comment by imre-kis-arm** [comment]
`CFG_THREAD_CORE_SHIFT` depends on the platform and it's hard to determine its value run time. There are platforms which have the MT bit set in their MPIDR registers but they are single threaded so on every processing element (core) the AFF0 is zero. Also there are multi threaded platforms like the Cortex-A65 where there are processing elements (threads) with 0 and 1 AFF0 values. In this case the `CFG_THREAD_CORE_SHIFT` should be 1 to avoid allocating same position value for multiple processing elements.
Basically the `CFG_THREAD_CORE_SHIFT` works the same way as the `CFG_CORE_CLUSTER_SHIFT` option. The platform's conf.mk should specify it or it should be passed in command line.
I used the term of "processing element" which is the generic way to name the unit which executes the instructions, so it can be a core or a thread depending on the platform.

File: core/arch/arm/kernel/misc_a32.S [hunk 1]
```diff
@@ -21,10 +22,30 @@ END_FUNC __get_core_pos
...
-	and	r0, r0, #MPIDR_CLUSTER_MASK
-	add	r0, r1, r0, LSR #(MPIDR_CLUSTER_SHIFT - CFG_CORE_CLUSTER_SHIFT)
+	mov	r3, r0
+
+	/*
+	 * Shift MPIDR value if it's not already shifted.
+	 * Using logical shift ensures AFF0 to be filled with zeroes.
+	 */
+	tst	r0, #MPIDR_MT_MASK
+	lsleq	r3, r0, #MPIDR_AFFINITY_BITS
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/misc_a32.S]: From the ARM ARM I can't tell that shift is needed. As I read it, the MT-bit is only a hint to the scheduler. Please clarify.

Comment (imre-kis-arm) [reply core/arch/arm/kernel/misc_a32.S]: There's some information about this in the ARM ARM D7.4 section's note (second point).
In practice a non-multi threaded platform's MPIDR values look like `(clusterId <<8) | coreId` and in multi threaded systems they follow the `MT | (clusterId << 16) | (coreId << 8) | threadId` formula. That shift operation is executed if the MT bit was 0 so the following instructions can work on a unified R0 value.

Comment (jenswi-linaro) [reply core/arch/arm/kernel/misc_a32.S]: Thanks for the pointer. I thought I answered here the other day, but it seems the message got lost somehow.

As I understand it `CFG_THREAD_CORE_SHIFT` must always be correctly defined to match how the CPU is configured so there's no need for semi-dynamic configuration. To keep the code simple and easier to read I would prefer something along:
``` C
#if CFG_THREAD_CORE_SHIFT == 0
	/* Calculate CorePos = (ClusterId * (cores/cluster)) + CoreId */
	and	r1, r0, #MPIDR_AFF0_MASK
	and	r0, r0, #MPIDR_AFF1_MASK
	add	r0, r1, r0, LSR #(MPIDR_AFF1_SHIFT - CFG_CORE_CLUSTER_SHIFT)
#else
	/*
	 * Calculate CorePos =
	 * ((ClusterId * (cores/cluster)) + CoreId) * (threads/core) + ThreadId
	 */
	ubfx	r2, r0, #MPIDR_AFF2_SHIFT, #MPIDR_AFFINITY_BITS
	ubfx	r1, r0, #MPIDR_AFF1_SHIFT, #MPIDR_AFFINITY_BITS
	ubfx	r0, r0, #MPIDR_AFF0_SHIFT, #MPIDR_AFFINITY_BITS
	add	r1, r1, r2, LSL #(CFG_CORE_CLUSTER_SHIFT)
	add	r0, r0, r1, LSL #(CFG_THREAD_CORE_SHIFT)
#endif
```
Perhaps and one pair of `ubfx` and `add` instructions can be combined, but it could be worth an extra instruction to keep it more readable.
I think we should use `CFG_CORE_THREAD_SHIFT` instead to harmonize better with all the other `CFG_` variables.
As an extra sanity check we could assert somewhere in `generic_boot.c` that with `CFG_THREAD_CORE_SHIFT` the MT bit isn't set and vice versa.

Comment (imre-kis-arm) [reply core/arch/arm/kernel/misc_a32.S]: I was a bit misleading in my previous comment because it's a valid setup for single threaded systems to have MPIDR values like this `MT | (clusterId << 16) | (coreId << 8) | 0` (threadId is 0 for all processing elements). So it's a valid case to have the MT bit set beside the CFG_THREAD_CORE_SHIFT is 0. However it's not valid to have the MT bit set, CFG_THREAD_CORE_SHIFT==0 but the AFF0 is **not** 0.
My point is that we need to shift the MPIDR by MPIDR_AFFINITY_BITS if the MT bit is zero to have the unified MPIDR layout before extracting the affinities and use this shifted value in the `#if CFG_THREAD_CORE_SHIFT == 0` block. We could still eliminate two unnecessary instructions. I hope this makes sense.
```c
	tst	r0, #MPIDR_MT_MASK
	lsleq	r3, r0, #MPIDR_AFFINITY_BITS

	/*
	 * At this point the MPIDR layout is always shifted so it looks
	 * as follows AFF2 -> cluster, AFF1 -> core, AFF0 -> thread
	 */
#if CFG_THREAD_CORE_SHIFT == 0
	/* Calculate CorePos = (ClusterId * (cores/cluster)) + CoreId */
	ubfx	r0, r3, #MPIDR_AFF1_SHIFT, #MPIDR_AFFINITY_BITS
	ubfx	r1, r3, #MPIDR_AFF2_SHIFT, #MPIDR_AFFINITY_BITS
	add	r0, r0, r1, LSL #(CFG_CORE_CLUSTER_SHIFT)
#else
	/*
	 * Calculate CorePos =
	 * ((ClusterId * (cores/cluster)) + CoreId) * (threads/core) + ThreadId
	 */
	ubfx	r0, r3, #MPIDR_AFF0_SHIFT, #MPIDR_AFFINITY_BITS
	ubfx	r1, r3, #MPIDR_AFF1_SHIFT, #MPIDR_AFFINITY_BITS
	ubfx	r2, r3, #MPIDR_AFF2_SHIFT, #MPIDR_AFFINITY_BITS
	add	r1, r1, r2, LSL #(CFG_CORE_CLUSTER_SHIFT)
	add	r0, r0, r1, LSL #(CFG_THREAD_CORE_SHIFT)
#endif
```

Comment (jenswi-linaro) [reply core/arch/arm/kernel/misc_a32.S]: Makes sense, please update the patch. An extra comment explaining what you said above about the MT bit would be nice.

It's likely that some new platforms will copy this function and modify it, so it's important that it's easy to understand how this function works.

**Comment by jenswi-linaro** [comment]
Please use `CFG_CORE_THREAD_SHIFT` instead of `CFG_THREAD_CORE_SHIFT`. `CFG_` variables which concern OP-TEE core only (S-EL1) are normally prefixed with `CFG_CORE_`. `CFG_CORE_CLUSTER_SHIFT` is one example.

While you're at it please squash the two commits together also and drop the `Change-Id:` tag.

To update you'll need to do a force push to this PR.

With that, please apply:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by imre-kis-arm** [comment]
Is there any scheduling for merging this pull request? I don't want to hurry you, I'm just asking.

**Comment by jforissier** [comment]
@imre-kis-arm although this is not a big change, I'd rather wait until after release 3.7.0 due in two days (Friday 18th) -- unless someone has a different opinion.

**Comment by imre-kis-arm** [comment]
It sounds good for me. Thank you.
